File: Sample Code/Borland Delphi 7/Simple PCSC/SimplePCSC.pas
////////////////////////////////////////////////////////////////////////////////
//
// Company  : FTsafe, LTD.
//
// Name     : SimplePCSC
//
////////////////////////////////////////////////////////////////////////////////

unit SimplePCSC;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, ExtCtrls, ComCtrls, FTsafeModule;

Const MAX_BUFFER_LEN    = 256;
Const INVALID_SW1SW2    = -450;

type
  TMainReadWrite = class(TForm)
    cbReader: TComboBox;
    mMsg: TRichEdit;
    bReader: TButton;
    bReset: TButton;
    GroupBox2: TGroupBox;
    bConnect: TButton;
    bStatus: TButton;
    bTransmit: TButton;
    bDisconnect: TButton;
    bRelease: TButton;
    gbData: TGroupBox;
    tDataIn: TEdit;
    gbOutData: TGroupBox;
    btStatus: TStatusBar;
    PollTimer: TTimer;
    bStartPolling: TButton;
    bStopPolling: TButton;
    procedure bReaderClick(Sender: TObject);
    procedure cbReaderChange(Sender: TObject);
    procedure FormActivate(Sender: TObject);
    procedure bResetClick(Sender: TObject);
    procedure bConnectClick(Sender: TObject);
    procedure bStatusClick(Sender: TObject);
    procedure bTransmitClick(Sender: TObject);
    procedure bDisconnectClick(Sender: TObject);
    procedure bReleaseClick(Sender: TObject);
    procedure tDataInKeyPress(Sender: TObject; var Key: Char);
    procedure PollTimerTimer(Sender: TObject);
    procedure bStartPollingClick(Sender: TObject);
    procedure bStopPollingClick(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  MainReadWrite: TMainReadWrite;
  hContext    : SCARDCONTEXT;
  hCard       : SCARDCONTEXT;
  ioRequest   : SCARD_IO_REQUEST;
  RdrState    : SCARD_READERSTATE;
  retCode     : Integer;
  dwActProtocol, BufferLen  : DWORD;
  SendBuff, RecvBuff        : array [0..262] of Byte;
  SendLen, RecvLen          : DWORD;
  Buffer      : array [0..MAX_BUFFER_LEN] of char;
  ConnActive, diFlag  : Boolean;

procedure ClearBuffers();
procedure InitMenu();
procedure DisableAPDUInput();
procedure displayOut(errType: Integer; retVal: Integer; PrintText: String);
function SendAPDU(): integer;
function TrimInput(dataInput: String): String;

implementation

{$R *.dfm}

procedure ClearBuffers();
var indx: integer;
begin

  for indx := 0 to 262 do
    begin
      SendBuff[indx] := $00;
      RecvBuff[indx] := $00;
    end;

end;

procedure InitMenu();
begin

  MainReadWrite.cbReader.Clear;
  MainReadWrite.mMsg.Clear;
  DisableAPDUInput();
  MainReadWrite.btStatus.SimpleText := 'Program ready';

end;

procedure DisableAPDUInput();
begin

  MainReadWrite.tDataIn.Clear;
  MainReadWrite.gbData.Enabled := False;

end;
procedure displayOut(errType: Integer; retVal: Integer; PrintText: String);
begin

  case errType of
    0: MainReadWrite.mMsg.SelAttributes.Color := clTeal;      // Notifications
    1: begin                                                  // Error Messages
         MainReadWrite.mMsg.SelAttributes.Color := clRed;
         PrintText := GetScardErrMsg(retVal);
       end;
    2: begin
         MainReadWrite.mMsg.SelAttributes.Color := clBlack;
         PrintText := '< ' + PrintText;                       // Input data
       end;
    3: begin
         MainReadWrite.mMsg.SelAttributes.Color := clBlack;
         PrintText := '> ' + PrintText;                       // Output data
       end;
    4: begin                                                  // Critical Error Messages
         MainReadWrite.mMsg.SelAttributes.Color := clRed;
         PrintText := PrintText;
       end;
  end;
  MainReadWrite.mMsg.Lines.Add(PrintText);
  MainReadWrite.mMsg.SelAttributes.Color := clBlack;

end;

function SendAPDU(): integer;
begin

  ioRequest.dwProtocol := dwActProtocol;
  ioRequest.cbPciLength := sizeof(SCARD_IO_REQUEST);
  retCode := SCardTransmit(hCard,
                           @ioRequest,
                           @SendBuff,
                           SendLen,
                           Nil,
                           @RecvBuff,
                           @RecvLen);
  if retCode <> SCARD_S_SUCCESS then begin
    DisplayOut(1, retCode, '');
    SendAPDU := retCode;
    Exit;
  end;
  SendAPDU := retCode;

end;

function TrimInput(dataInput: String): String;
var indx: Integer;
    tmpStr: String;
begin
  tmpStr := '';
  dataInput := Trim(dataInput);
  for indx := 1 to Length(dataInput) do
    if dataInput[indx] <> chr(32) then
      tmpStr := tmpStr + dataInput[indx];
  diFlag := False;
  if (Length(tmpStr) mod 2) <> 0 then
    diFlag := True;
  TrimInput := tmpStr;
  
end;

procedure TMainReadWrite.bReaderClick(Sender: TObject);
begin

  // 1. Establish context and obtain hContext handle
  retCode := SCardEstablishContext(SCARD_SCOPE_USER,
                                   nil,
                                   nil,
                                   @hContext);
  if retCode <> SCARD_S_SUCCESS then
    begin
      DisplayOut(1, retCode, '');
      Exit;
    end
  else
    begin
      DisplayOut(0, 0, 'SCardEstablishContext... OK');
    end;

  // 2. List PC/SC card readers installed in the system
  BufferLen := MAX_BUFFER_LEN;
  retCode := SCardListReadersA(hContext,
                               nil,
                               @Buffer,
                               @BufferLen);
  if retCode <> SCARD_S_SUCCESS then
    begin
      DisplayOut(1, retCode, '');
      Exit;
    end
  else
    DisplayOut(0, 0, 'SCardListReaders... OK');

  MainReadWrite.cbReader.Clear;
  LoadListToControl(MainReadWrite.cbReader,@buffer,bufferLen);
  MainReadWrite.cbReader.ItemIndex := 0;
  bConnect.Enabled := True;
  bRelease.Enabled := True;
  bStartPolling.Enabled := True ;
  bStopPolling.Enabled := True ;
  bReader.Enabled := False;

end;

procedure TMainReadWrite.cbReaderChange(Sender: TObject);
begin


  if ConnActive then
  begin
    retCode := SCardDisconnect(hCard, SCARD_UNPOWER_CARD);
    ConnActive := False;
  end;

end;

procedure TMainReadWrite.FormActivate(Sender: TObject);
begin

  InitMenu();

end;


procedure TMainReadWrite.bResetClick(Sender: TObject);
begin
    MainReadWrite.mMsg.Clear;
end;

procedure TMainReadWrite.bConnectClick(Sender: TObject);
begin
   if ConnActive then
  begin
    DisplayOut(0, 0,'Connection is already active.');
    Exit;
  end;

  // 1. Connect to selected reader using hContext handle
  //    and obtain valid hCard handle
  retCode := SCardConnectA(hContext,
                           PChar(cbReader.Text),
                           SCARD_SHARE_SHARED,
                           SCARD_PROTOCOL_T0 or SCARD_PROTOCOL_T1,
                           @hCard,
                           @dwActProtocol);
  if retCode <> SCARD_S_SUCCESS then
    begin
      DisplayOut(1,retCode, '');
      ConnActive := False;
      Exit;
    end
  else
    DisplayOut(0, 0,'SCardConnect...OK');

  ConnActive := True;
  bDisConnect.Enabled := True;
  bRelease.Enabled := True;
  gbData.Enabled := True;
  tDataIn.SetFocus;
  bStatus.Enabled := True;
  bTransmit.Enabled := True;

  bReader.Enabled := False;
  bConnect.Enabled := False;
  bStartPolling.Enabled := False ;
  bStopPolling.Enabled := False ;

end;

procedure TMainReadWrite.bStatusClick(Sender: TObject);
var ReaderLen, dwState: Integer;
    tmpStr: String;
    ATRVal: array [0..128] of byte;
    indx: Integer;
    ATRLen: DWORD;
begin

  // Invoke SCardStatus using hCard handle
  //    and valid reader name
  ATRLen := 32;
  ReaderLen := 0;
  dwState := 0;
  retCode := SCardStatusA(hCard,
                         PChar(cbReader.Text),
                         @ReaderLen,
                         @dwState,
                         @dwActProtocol,
                         @ATRVal,
                         @ATRLen);  //@ATRLen);
  if retCode <> SCARD_S_SUCCESS then
    begin
      DisplayOut(1,retCode, '');
      ConnActive := False;
      Exit;
    end
  else
    DisplayOut(0, 0, 'SCardStatus...OK');

  // Format ATRVal returned and display string as ATR value
  tmpStr := '';
  for indx := 0 to ATRLen-1 do
    begin
      tmpStr := tmpStr + Format('%.02X ',[ATRVal[indx]]);
    end;
  DisplayOut(3, 0,Format('ATR Value: %s',[tmpStr]));

    // Interpret dwState returned and display as active protocol
  tmpStr := '';
  case integer(dwState) of
    0: tmpStr := 'SCARD_UNKNOWN';
    1: tmpStr := 'SCARD_ABSENT';
    2: tmpStr := 'SCARD_PRESENT';
    3: tmpStr := 'SCARD_SWALLOWED';
    4: tmpStr := 'SCARD_POWERED';
    5: tmpStr := 'SCARD_NEGOTIABLE';
    6: tmpStr := 'SCARD_SPECIFIC';
    else
      tmpStr := '';
    end;
  DisplayOut(3, 0,Format('Reader State: %s',[tmpStr]));

  // Interpret dwActProtocol returned and display as active protocol
  tmpStr := '';
  case integer(dwActProtocol) of
    1: tmpStr := 'T=0';
    2: tmpStr := 'T=1';
    else
      tmpStr := 'No protocol is defined.';
    end;
  DisplayOut(3, 0,Format('Active Protocol: %s',[tmpStr]));
  ATRLen := 0;

end;

procedure TMainReadWrite.bTransmitClick(Sender: TObject);
var tmpStr, sw1sw2: String;
    indx: Integer;
begin

  if tDataIn.Text = '' then begin
    DisplayOut(4, 0, 'No data input');
    tDataIn.SetFocus;
    Exit;
  end;
  tmpStr := TrimInput(tDataIn.Text);
  if Length(tmpStr) <10 then begin
    DisplayOut(4, 0, 'Insufficient data input');
    tDataIn.SetFocus;
    Exit;
  end;
  if diFlag then begin
    DisplayOut(4, 0, 'Invalid data input, uneven number of characters');
    tDataIn.SetFocus;
    Exit;
  end;
  ClearBuffers();
  For indx := 0 to 4 do
    SendBuff[indx] := StrToInt('$' + Copy(tmpStr, indx*2+1, 2));

  // if APDU length < 6 then P3 is Le
  if Length(tmpStr) <12 then begin
    For indx := 0 to 4 do
      SendBuff[indx] := StrToInt('$' + Copy(tmpStr, indx*2+1, 2));
    SendLen := $05;
    RecvLen := SendBuff[4] + 2;
    tmpStr := '';
    for indx := 0 to 4 do
      tmpStr := tmpStr + Format('%.02X ',[SendBuff[indx]]);
    DisplayOut(2, 0, tmpStr);
    retCode := SendAPDU();
    if retCode = SCARD_S_SUCCESS then begin
      tmpStr :='';
      for indx := 0 to RecvLen-1 do
        tmpStr := tmpStr + Format('%.02X ',[RecvBuff[indx]]);
      DisplayOut(3, 0, tmpStr);
    end;
  end
  else begin
    // P3 is Lc
    For indx := 0 to 4 do
      SendBuff[indx] := StrToInt('$' + Copy(tmpStr, indx*2+1, 2));
    SendLen := $05 + SendBuff[4];
    if Length(tmpStr) <  SendLen*2 then begin
      DisplayOut(4, 0, 'Invalid data input, insufficient data length');
      tDataIn.SetFocus;
      Exit;
    end;
    For indx := 0 to SendBuff[4]-1 do
      SendBuff[indx+5] := StrToInt('$' + Copy(tmpStr, (indx+5)*2+1, 2));
    RecvLen := $02;
    tmpStr := '';
    for indx := 0 to SendLen-1 do
      tmpStr := tmpStr + Format('%.02X ',[SendBuff[indx]]);
    DisplayOut(2, 0, tmpStr);
    retCode := SendAPDU();
    if retCode = SCARD_S_SUCCESS then begin
      tmpStr :='';
      for indx := 0 to RecvLen-1 do
        tmpStr := tmpStr + Format('%.02X ',[RecvBuff[indx]]);
      DisplayOut(3, 0, tmpStr);
    end;
  end;

end;


procedure TMainReadWrite.bDisconnectClick(Sender: TObject);
begin
  If ConnActive Then
    retCode := SCardDisconnect(hCard, SCARD_UNPOWER_CARD);
    If retCode <> SCARD_S_SUCCESS Then
    begin
        DisplayOut(1, retCode, '');
        Exit;
    end
    else
      DisplayOut(0, 0, 'SCardDisconnect...OK');
  ConnActive := False;
  bConnect.Enabled := True;
  bReader.Enabled := True;
  bRelease.Enabled := True;
  bStartPolling.Enabled := True ;
  bStopPolling.Enabled := True ;
  bDisconnect.Enabled := False;
  bTransmit.Enabled := False;
  bStatus.Enabled := False;
end;

procedure TMainReadWrite.bReleaseClick(Sender: TObject);
begin
    If ConnActive Then
     retCode := SCardReleaseContext(hContext);

    If retCode <> SCARD_S_SUCCESS Then
    begin
        DisplayOut(1, retCode, '');
        Exit;
    end
    else
      DisplayOut(0, 0, 'SCardReleaseContext...OK');
      
    cbReader.Items.Clear;
    cbReader.Refresh;
    cbReader.ItemIndex := 0;
    bRelease.Enabled := False;
    bReader.Enabled := True;
    bConnect.Enabled := False;
    bStartPolling.Enabled := False;
    bStopPolling.Enabled := False;
    bTransmit.Enabled := False;
    bStatus.Enabled := False;
    bDisconnect.Enabled := False;

end;

procedure TMainReadWrite.tDataInKeyPress(Sender: TObject; var Key: Char);
begin
  if Key in ['a'..'z'] then Dec(Key,32);
  if Not(Key in ['0'..'9', 'A'..'F', chr(8), chr(32)]) then
    Key := #0;
end;

procedure TMainReadWrite.PollTimerTimer(Sender: TObject);
begin
  MainReadWrite.mMsg.Text :='';
  RdrState.szReader := PChar(MainReadWrite.cbReader.Text);
  retCode := SCardGetStatusChangeA(hContext,
                                 0,
                                 @RdrState,
                                 1);
  if retCode <> SCARD_S_SUCCESS then
    begin
      DisplayOut(1, retCode, '');
      MainReadWrite.PollTimer.Enabled := False;
      Exit;
    end ;
  if (RdrState.dwEventStates and SCARD_STATE_PRESENT) <> 0 then
    DisplayOut(0, 0, 'Card is inserted')
  else
    DisplayOut(0, 0, 'Card is removed');
    
end;

procedure TMainReadWrite.bStartPollingClick(Sender: TObject);
begin
    bStartPolling.Enabled := false;
    bConnect.Enabled := False;
    PollTimer.Enabled := True ;
end;

procedure TMainReadWrite.bStopPollingClick(Sender: TObject);
begin
    bStartPolling.Enabled := True;
    bConnect.Enabled := True;
    PollTimer.Enabled := False ;
    DisplayOut(0, 0, 'StopPolling...OK');
end;

end.

File: Sample Code/MS Visual C#/Simple_PCSC/Simple_PCSC/Simple_PCSC.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace Simple_PCSC
{
    public partial class Simple_PCSC : Form
    {
        uint PRT;
        int ATRLen, ReaderLen, dwState;
        byte[] ATR = new byte[33];
        SCARD_IO_REQUEST IORequest;
        public int hCard, hContext;                     // hContext=10
        public int Protocol, ReaderCount;
        public string CardReaderDesc = "";              // Save the device descriptor string

        public Simple_PCSC()
        {
            InitializeComponent();
        }


        // Byte[] to HexString
        static char[] hexDigits = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
        public static string ToHexString(byte[] bytes, int BytesLength)
        {
            char[] chars = new char[bytes.Length * 5];
            for (int i = 0; i < BytesLength; i++)
            {
                int b = bytes[i];
                chars[i * 3 + 0] = hexDigits[b >> 4];
                chars[i * 3 + 1] = hexDigits[b & 0xF];
                chars[i * 3 + 2] = ' ';
            }
            return new string(chars);
        }

        // HexString to Byte[]
        private static byte[] ToHexByte(string hexString)
        {
            hexString = hexString.Replace(" ", "");
            if ((hexString.Length % 2) != 0)
                hexString += " ";
            byte[] returnBytes = new byte[(hexString.Length / 2)];
            for (int i = 0; i < returnBytes.Length; i++)
                returnBytes[i] = Convert.ToByte(hexString.Substring(i * 2, 2), 16);

            return returnBytes;
        }

        void DisplayOut(int mType, long msgCode, string PrintText)
        {
            switch (mType)
            {
                case 0:                                  // Notifications only
                    mMsg.SelectionColor = Color.Green;
                    break;
                case 1:                                  // Error Messages
                    mMsg.SelectionColor = Color.Red;
                    PrintText = ModWinsCard.GetScardErrMsg((int)msgCode);
                    break;
                case 2:                                  // Input data
                    mMsg.SelectionColor = Color.Black;
                    PrintText = "< " + PrintText;
                    break;
                case 3:                                  // Output data
                    mMsg.SelectionColor = Color.Black;
                    PrintText = "> " + PrintText;
                    break;
                case 4:                                  // Critical Errors
                    mMsg.SelectionColor = Color.Red;
                    break;
            }

            mMsg.SelectedText = PrintText + "\n";
            mMsg.SelectionStart = mMsg.Text.Length;
            mMsg.SelectionColor = Color.Black;
        }

        private bool ListCardReaderDesc()
        {
            int RetListCardReader;
            int PccHReaders = 0;


            //Establish Context
            RetListCardReader = ModWinsCard.SCardEstablishContext(ModWinsCard.SCARD_SCOPE_USER, 0, 0, ref hContext);

            if (RetListCardReader != ModWinsCard.SCARD_S_SUCCESS)
            {
                return false;
            }

            // 2. List PC/SC card readers installed in the system
            RetListCardReader = ModWinsCard.SCardListReaders(hContext, null, null, ref PccHReaders);

            if (RetListCardReader != ModWinsCard.SCARD_S_SUCCESS)
            {
                return false;
            }

            string ReaderList = "" + Convert.ToChar(0);
            int Index;
            string ReaderName = "";
            ReaderName = "";
            Index = 0;
            PccHReaders = 255;
            byte[] ReadersList = new byte[PccHReaders];

            // Fill reader list
            RetListCardReader = ModWinsCard.SCardListReaders(hContext, null, ReadersList, ref PccHReaders);

            if (RetListCardReader != ModWinsCard.SCARD_S_SUCCESS)
            {
                return false;
            }
            if (CardReaderDesc == "")
            {
                ReaderListComboBox.Items.Clear();
            }
            //Convert reader buffer to string
            while (ReadersList[Index] != 0)
            {

                while (ReadersList[Index] != 0)
                {
                    ReaderName = ReaderName + (char)ReadersList[Index];
                    Index = Index + 1;
                }

                if (CardReaderDesc == "")
                {
                    //Add reader name to list
                    //  if (string.Compare(ReaderName, bR301_ReaderName) == 0)
                    {
                        ReaderListComboBox.Items.Add(ReaderName);
                    }
                }
                else
                {
                    if (string.Compare(ReaderName, CardReaderDesc) == 0)
                    {
                        return true;
                    }
                    else
                    {
                        ReaderListComboBox.Items.Clear();
                        ReaderListComboBox.Items.Add(ReaderName);
                        return true;
                    }
                }
                ReaderName = "";
                Index = Index + 1;
            }
            return true;
        }

        private void ReaderInitButton_Click(object sender, EventArgs e)
        {
            ReaderListComboBox.Items.Clear();
            ReaderListComboBox.Text = "";
            CardReaderDesc = "";


            if (ListCardReaderDesc())
            {
                ReaderListComboBox.SelectedIndex = 0;
                DisplayOut(0, 0, "SCardListReaders... OK");
                ReaderConnectButton.Enabled = true;
                StartPpollingButton.Enabled = true;
                StopPpollingButton.Enabled = true;

                ReaderReleaseContextbutton.Enabled = true;
            }
            else
            {
                DisplayOut(4, 0, "SCardListReaders... ERR");
            }
            
        }

        private void ReaderConnectButton_Click(object sender, EventArgs e)
        {
            CardReaderDesc = ReaderListComboBox.SelectedItem.ToString();

            // Don't use SCARD_SHARE_DIRECT , After Ppolling the transmit return 0x00000016 Unknown Error
            int retCode = ModWinsCard.SCardConnect(this.hContext, CardReaderDesc, ModWinsCard.SCARD_SHARE_SHARED,
                                                    ModWinsCard.SCARD_PROTOCOL_T0 | ModWinsCard.SCARD_PROTOCOL_T1, ref hCard, ref Protocol);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
                DisplayOut(1, retCode, "");

            else
            {
                DisplayOut(0, 0, "SCardConnect...OK");
                ReaderConnectButton.Enabled = false;
                StartPpollingButton.Enabled = false;
                StopPpollingButton.Enabled = false;

                ReaderTransmitButton.Enabled = true;
                ReaderDisconnectButton.Enabled = true;
                ReaderReleaseContextbutton.Enabled = true;
                ReaderStatusButton.Enabled = true;
            }
        }

        private void ReaderTransmitButton_Click(object sender, EventArgs e)
        {
            if (m_Apdu.Text == "")
            {
                DisplayOut(4, 0, "No data input");
                return;
            }

            m_Apdu.Text = (m_Apdu.Text.Replace(" ", ""));
            if (m_Apdu.Text.Length < 10)
            {
                DisplayOut(4, 0, "Insufficient data input");
                return;
            }

            if ((m_Apdu.Text.Length % 2) != 0)
            {
                DisplayOut(4, 0, "Invalid data input, uneven number of characters");
                m_Apdu.Focus();
                return;
            }

            byte[] SendBuff = new byte[1024];
            byte[] RecvBuff = new byte[1024];
            int SendLen = 1024;
            int RecvLen = 1024;

            SendLen = m_Apdu.Text.Length / 2;
            byte[] tempBuff = ToHexByte(m_Apdu.Text);
            Array.Copy(tempBuff, SendBuff, SendLen);

            IORequest.dwProtocol = Protocol;
            IORequest.cbPciLength = 8;
            int retCode = ModWinsCard.SCardTransmit(hCard, ref IORequest, ref SendBuff[0], SendLen, ref IORequest, ref RecvBuff[0], ref RecvLen);
            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {
                DisplayOut(1, retCode, "");
            }
            else
            {
                string tempStr = ToHexString(RecvBuff, RecvLen);
                DisplayOut(0, 0, tempStr);
            }
        }

        private void ReaderDisconnectButton_Click(object sender, EventArgs e)
        {
            int retCode = ModWinsCard.SCardDisconnect(hCard, ModWinsCard.SCARD_UNPOWER_CARD);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
                DisplayOut(1, retCode, "");
            else
            {
                DisplayOut(0, 0, "SCardDisconnect...OK");
                ReaderConnectButton.Enabled = true;
                StartPpollingButton.Enabled = true;
                StopPpollingButton.Enabled = true;

                ReaderStatusButton.Enabled = false;
                ReaderTransmitButton.Enabled = false;
                ReaderDisconnectButton.Enabled = false;
            }
        }

        private void ReaderReleaseContextbutton_Click(object sender, EventArgs e)
        {
            CardReaderDesc = "";
            int retCode = ModWinsCard.SCardReleaseContext(hContext);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
                DisplayOut(1, retCode, "");
            else
            {
                DisplayOut(0, 0, "SCardReleaseContext...OK");

                ReaderConnectButton.Enabled = false;
                StartPpollingButton.Enabled = false;
                StopPpollingButton.Enabled = false;

                ReaderTransmitButton.Enabled = false;
                ReaderDisconnectButton.Enabled = false;
                ReaderReleaseContextbutton.Enabled = false;
                ReaderStatusButton.Enabled = false;
            }
        }

        private void timer1_Tick(object sender, EventArgs e)
        {
            // perform if the card is inserted/not inserted on the reader 

            // Don't connection before to connect card reader.
            // If the connection card reader to make a card, and after make smart card device manager to suspend work, and can't get correct card slots.
            SCARD_READERSTATE ReaderState = new SCARD_READERSTATE();
            ReaderState.RdrName = ReaderListComboBox.SelectedItem.ToString();
            
            mMsg.Clear();

            int retCode = ModWinsCard.SCardGetStatusChange(this.hContext, -1, ref ReaderState, 1);
            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {
                DisplayOut(1, retCode, "");
            }
            else
            {
                if ((ReaderState.RdrEventState & ModWinsCard.SCARD_STATE_PRESENT) != 0)
                {
                    DisplayOut(0, 0, "Card Inserted");
                }
                else
                {
                    DisplayOut(0, 0, "Card Removed");
                }
            }
        }

        private void StartPpollingButton_Click(object sender, EventArgs e)
        {
            timer1.Enabled = true;
            ReaderConnectButton.Enabled = false;
            StartPpollingButton.Enabled = false;
            ReaderReleaseContextbutton.Enabled = false;
        }

        private void StopPpollingButton_Click(object sender, EventArgs e)
        {
            StartPpollingButton.Enabled = true;
            ReaderConnectButton.Enabled = true;
            ReaderReleaseContextbutton.Enabled = true;
            timer1.Enabled = false;
            DisplayOut(0, 0, "Stop Ppolling");
        }

        private void button1_Click(object sender, EventArgs e)
        {
            short index;
            string Temp;
            ATRLen = 33;
            CardReaderDesc = ReaderListComboBox.SelectedItem.ToString();

            //perform the Card Status
            int retcode = ModWinsCard.SCardStatus(this.hCard, CardReaderDesc, ref ReaderLen, ref dwState, ref Protocol, ref ATR[0], ref ATRLen);

            if (retcode != ModWinsCard.SCARD_S_SUCCESS)
            {
                DisplayOut(1, retcode, "");
            }
            else
            {
                DisplayOut(0, 0, "SCardStatus OK...");

                // select for the protocol
                switch (Protocol)
                {
                    case 0: PRT = ModWinsCard.SCARD_PROTOCOL_UNDEFINED;
                        DisplayOut(0, 0, "Active Protocol Undefined");
                        break;
                    case 1: PRT = ModWinsCard.SCARD_PROTOCOL_T0;
                        DisplayOut(0, 0, "Active Protocol T0");
                        break;
                    case 2: PRT = ModWinsCard.SCARD_PROTOCOL_T1;
                        DisplayOut(0, 0, "Active Protocol T1");
                        break;
                }

                Temp = "ATR:";

                for (index = 0; index < ATRLen; index++)
                    Temp = Temp + " " + string.Format("{0:X2}", ATR[index]);

                // Display ATR value 
                DisplayOut(0, 0, Temp);
            }
        }

        private void button1_Click_1(object sender, EventArgs e)
        {
            mMsg.Clear();
        }

    }
}
